### ISSUE TITLE
Doubts about the implementation.

### ISSUE THREAD

**Issue by GibranLara** [andre-artus/kotlin-optional-param-spring-demo#1]
**Error running the application.**
- When the application is executed, the connection is opened and closed, so the application is not executing. I'm using IntelliJ 2007.5.2.

```
2018-11-14 18:33:21.970  INFO 16216 --- [           main] org.mongodb.driver.connection            : Opened connection [connectionId{localValue:2, serverValue:239}] to localhost:27017
2018-11-14 18:33:22.016  INFO 16216 --- [      Thread-16] org.mongodb.driver.connection            : Closed connection [connectionId{localValue:2, serverValue:239}] to localhost:27017 because the pool has been closed.

Process finished with exit code 0
```

**Domain classes** 
- The domain classes 'Meeting' and 'Project', their role in maintaining data, but only those that have already been filtered and validated?

- Are the domain classes the same as those used as models or entities?

- The annotation @Valid what is its function? I understand that it validates the model that receives the function, but in itself, what does it validate?

- The classes MeetingRequest and ProjectRequest, how are they called, I mean if they are model type classes, entity, domain?

- What are the modules that work in the business logic layer? I understood that only the controllers. And that both repositories and services were part of the data access layer.

**Comment by andre-artus** [comment]
What do you get when you run  `.\gradlew bootRun` on the command line (from the `kotlin-spring-demo` directory)?

If it exists without showing a line `Started KotlinSpringDemoApplicationKt in ...` then pass the output through to a text file and upload it here:
(like this)
`.\gradlew bootRun > output.txt`

If that command does not return back to the prompt then that means the app is running.

**Comment by andre-artus** [comment]
`localhost:27017` is the default MongoDB host, you may need to change that according to your own setup.

**Comment by GibranLara** [comment]
> .\gradlew bootRun

Yes, if run from command line with .\gradlew bootRun the app run correctly.

**Comment by andre-artus** [comment]
> **Domain classes**
> 
>     * The domain classes 'Meeting' and 'Project', their role in maintaining data, but only those that have already been filtered and validated?
> 
Yes, those represent the internal state of your application. This is the data that the app manipulates to perform it's use cases.

>     * Are the domain classes the same as those used as models or entities?
> 
Yes, sometimes. There is often levels. You can write an app that has no knowledge of how the data is stored or received, so long as it conforms to an agreed upon interface. You can have an internal model for your web api, a request and response model, a different internal model for your mobile app, etc. An entity is just a chunk of data that is uniquely identified. The annotations, as used by Spring, just helps with some code gen and wiring up.

>     * The annotation @Valid what is its function? I understand that it validates the model that receives the function, but in itself, what does it validate?
> 
That's a leftover that I forgot to remove, it's related to Spring's built in validation features. It's not used in this example. I was trying to keep things simple, and in line with what you already have. Although, as I said over on SO, this is not the correct way to do REST services. If I have time I'll see about knocking something together.

>     * The classes MeetingRequest and ProjectRequest, how are they called, I mean if they are model type classes, entity, domain?
> 
They are part of the IO model, [Class]Request is for input, [Class]Response is for output. It's usually not a good idea to use tour internal model for IO. I'll have to explain in more detail when I have time. Think of it like connectors for a phone, computer, or TV. 

If I have [simple] USB device, I can plug it into any USB port and get it working without having to open up the case and re-solder connectors (unless it an Apple device, of course).

You want to be able to keep your IO model relatively stable while still being able to make major changes to the internals of your app.

 If you swap a graphics card on your computer all your USB devices should not suddenly stop working.

>     * What are the modules that work in the business logic layer? I understood that only the controllers. And that both repositories and services were part of the data access layer.
Controllers are actually part of the IO. You would ideally have a different set of coordinating classes called "Use Cases" that perform the business functions.

What development have you done? Just so I know how much detail I need to go into.

**Comment by andre-artus** [comment]
> 
> 
> > .\gradlew bootRun
> 
> Yes, if run from command line with .\gradlew bootRun the app run correctly.

Can you update IDEA?

**Comment by andre-artus** [comment]
Intellij IDEA 2007.5.2 is VERY old. Are you on the Ultimate version?

**Comment by andre-artus** [comment]
I have the Jetbrains toolbox that lets me switch between different versions, and that only goes back so far as 2016.

**Comment by GibranLara** [comment]
> Intellij IDEA 2007.5.2 is VERY old. Are you on the Ultimate version?

Sorry I mean 2017.5.2.

**Comment by GibranLara** [comment]
> What development have you done? Just so I know how much detail I need to go into.

I have worked with Grails 2.5.1, I made a small system. But on the subject of API's REST I am completely a rookie. In grails I worked with a structure of services, domains and controllers. So I have never exposed data to a front end.

When I worked with Grails I was with the idea that it is of the MVC pattern.
The model (M) was all part of domain classes, repositories, services and data access. The controllers were the controller. And finally, the view was the GSP.

But now I am confused by what you told me about the layers.

Regarding the Multitier architecture. I thought it was the next order.

Presentation layer (HTML)
Application layer (Domain classes, entities or models)
Business layer (Controllers)
Data access layer (Repositories, services).

And I think I'm very wrong.

**Comment by GibranLara** [comment]
> Intellij IDEA 2007.5.2 is VERY old. Are you on the Ultimate version?

I'm installing 2018.2.6.

**Comment by andre-artus** [comment]
> 
> 
> > Intellij IDEA 2007.5.2 is VERY old. Are you on the Ultimate version?
> 
> I'm installing 2018.2.6.

Let me know how you get on, I'm going to sleep.

**Comment by GibranLara** [comment]
> > > Intellij IDEA 2007.5.2 is VERY old. Are you on the Ultimate version?
> > 
> > 
> > I'm installing 2018.2.6.
> 
> Let me know how you get on, I'm going to sleep.

Good night. I really appreciate and thank you so much that you are taking your time to explain, thank you.

**Comment by andre-artus** [comment]
Do yourself a favour, go watch this:
https://www.youtube.com/watch?v=WpkDN78P884
While it's given at a Ruby conference it's not specific to either Ruby or Rails. It discusses general architecture best practices.
You can also search for "Clean Architecture".

Don't be tempted to go seek the "perfect" version, as that depends very much on context.

Most important thing: your application logic should be independent of how it does IO (UI, DB, etc.). You should be able to package it into a separate library (e.g. Jar). In fact that is a good idea generally as it can help speed up compilation. But the real reasons are: test-ability and isolation/independence. With my applications I often decide to just plug in a scripting interface as IO (on what could otherwise be a mobile or web app) so that I can automate some action from the command line or script file. It goes through all the input validation and rules as the primary user facing UI. This does not take a lot of work because all the real logic is in the library. Being able to do that also means you can easily make a web app into a mobile app, or desktop UI, or even a headless service.

When reading or viewing media about software architecture or patterns it is important to keep in mind that there is often a lot of overloading of meaning (e.g double meaning) between terms so it is easy to get confused. As an example in the video above MVC has an original [Trygve Reenskaug] meaning (e.g. smalltalk and cocoa)  and a web meaning (really "Model 2") that originated in J2EE but was popularized by Rails. Or, as you asked earlier, "Entity" can have different meaning depending on context.

**Comment by GibranLara** [comment]
> Do yourself a favour, go watch this:

Of course I will, thank you.

Yesterday you also told me that I was performing the construction of the Rest API incorrectly, what are the differences that exist with web MVC and what should I work with to modify it and be a correct Rest API?

**Comment by andre-artus** [comment]
> Yesterday you also told me that I was performing the construction of the Rest API equitably, 

I'm not sure if this is a translation error, but I'm pretty sure that was not my intention. Did you mean "incorrectly"?

> what are the differences that exist with web MVC and what should I work with to modify it and be a correct Rest API?

Look into:

- `Spring Rest Repositories `, and
- `Spring HATEOAS` 

Those are two out of the box options. As is normally the problem with out-of-the-box solutions they are tied to the framework. But play with those to get the idea. Each should go into their own project (conceptually competing).

If you look at the following, the response handling --especially error handling-- is not very good for an REST API.

```
 fun createValidProject(@RequestBody projectRequest: ProjectRequest): ResponseEntity<Project> {
        return try {
            val validProject: Project = projectRequest.validate()
            // Should be ResponseEntity.created()
            ResponseEntity.ok(projectService.createProject(validProject))
        } catch (e: IllegalArgumentException) {
            logger.error(e.localizedMessage)
            // Should be something like ResponseEntity.badRequest().body("Blah : ${e.localizedMessage ?: e.message}")
            ResponseEntity.badRequest().build()
        }
    }
```
### Web
The typical way the **Web** MVC framework expects to work is that an error either redirects to the same page (input validation error, e.g. 400) or to an error page (404, etc.), or a login page (e.g. 401). That is reflected int the return type (ResponseEntity). 

Web UI interactions are usually limited to GET and POST due to historical browser limitations. The Spring MVC Web framework does a half-baked attempt at REST. `ResponseEntity.created(location)` expects a redirect address (`location`), and while `ResponseEntity.ok` will return an object, it does so with a HTTP `200 OK` code (as the name suggests). 
The error messages are better suited for web pages.

### API
A REST **API** (which may __or may not__ follow MVC) :
* The request does not redirect (other than perhaps for select 3xx responses), redirection is handled by the client.
* It uses the HTTP verbs to indicate intent (GET, PUT, DELETE, POST, etc.).
* Returns HTTP status codes (`201 Created`, `202`, etc.) along with an error message (when applicable) in the same format as indicated in the `Accept` header.

There are of course other differences. But they can be conflated in the framework (as is often done by Spring [and others]).

### General
There are a lot of decisions made in Spring [Boot] that can quickly get a project on its feet, but that becomes questionable when trying to build something more long-lived. Like the use of `PagingAndSortingRepository` and its children (e.g. `MongoRepository`). The repository pattern is actually intended to uncouple your data store from the application, using this interface couples it. Also the use of annotations like `@Entity` confuses people about where an class/object should reside.

This project of yours, is it for work or for learning?

**Comment by GibranLara** [comment]
> "incorrectly"

Yes, that was the word, I try to translate as best as possible, but I miss some words, excuse me.

> ### General
> There are a lot of decisions made in Spring [Boot] that can quickly get a project on its feet, but that becomes questionable when trying to build something more long-lived. Like the use of `PagingAndSortingRepository` and its children (e.g. `MongoRepository`). The repository pattern is actually intended to uncouple your data store from the application, using this interface couples it. Also the use of annotations like `@Entity` confuses people about where an class/object should reside.

Now I understand too many things.



> This project of yours, is it for work or for learning?

It is an exercise to learn. I wanted to make a small compilation of technologies to learn the basics of each one.

Backend.

- Build a REST API with Spring Boot.
- Programming with Kotlin.
- Store data in MongoDB.

Frontend.

- Use Vue and Vuetify as a framework.

But just beginning I found myself with my problem of how to generate an object id for an embedded document. : D

**Comment by andre-artus** [comment]
> > "incorrectly"
> 
> Yes, that was the word, I try to translate as best as possible, but I miss some words, excuse me.

No problem, I'm also a non-native English speaker. Feel free to ask me to clarify if I'm writing in a way that is difficult to understand.

> 
> Now I understand too many things.

Is it possible to understand too many things, or am I missing something?

> > This project of yours, is it for work or for learning?
> 
> It is an exercise to learn. I wanted to make a small compilation of technologies to learn the basics of each one.
 
That's cool. So you would not mind that we discuss the details of it or collaborate in public.

> Backend.
> 
>     * Build a REST API with Spring Boot.
> 
>     * Programming with Kotlin.
> 
>     * Store data in MongoDB.
> 
> 
> Frontend.
> 
>     * Use Vue and Vuetify as a framework.

That is a very good set, I don;t know how it is over where you live, but over here that skill-set can add 10-20% to your income (if you want to work in corporate). 

> But just beginning I found myself with my problem of how to generate an object id for an embedded document. : D

By the way, you should consider editing your SO question to make it clear to other users. It made sense to me, mostly because of my experience, but it obviously confusing to others given the comments and rating.

There are actually a few more issues with the design that I did not want to go into over at SO.
 I left out the participants from this example for 2 reasons:

1. to keep it simple,
2. It seems broken design-wise.

**Comment by GibranLara** [comment]
> Is it possible to understand too many things, or am I missing something?

I mean that with the explanation you gave of Web, Api and General, many things made more sense to me. I understood and several concepts.


> That's cool. So you would not mind that we discuss the details of it or collaborate in public.

I'm sorry, I did not understand this point.


> That is a very good set, I don;t know how it is over where you live, but over here that skill-set can add 10-20% to your income (if you want to work in corporate).

More than anything I do it to try different technologies. It is good to hear that they are well paid, although I do not know if it is the same in Mexico.

> By the way, you should consider editing your SO question to make it clear to other users.

All right, I'll change the title of the question.

**Comment by andre-artus** [comment]
> I mean that with the explanation you gave of Web, Api and General, many things made more sense to me. I understood and several concepts.
> 

Cool. Glad that helped. I find that idioms do not translate well.

> > That's cool. So you would not mind that we discuss the details of it or collaborate in public.
> 
> I'm sorry, I did not understand this point.
 
What I mean is that this GitHub repository (again a overloaded word) is viewable by other people (the public). If it was a work application there would probably be restrictions on discussing the internal working of it in public, but as it is a learning project, you probably don't mind public discussion.

> > That is a very good set, I don;t know how it is over where you live, but over here that skill-set can add 10-20% to your income (if you want to work in corporate).
> 
> More than anything I do it to try different technologies. It is good to hear that they are well paid, although I do not know if it is the same in Mexico.
 
Just compare job listings. There are some rules of the IT game when it comes to compensation:
* The more you need to fight the framework to get useful things done, the higher the compensation.
* If the framework is sold to managers by consultants (sales people), on a golf course, the higher the compensation.
* If it is in the news or has some buzz (like Rails a few years back). Higher compensation.
* If the framework is nice to work with, is loved by programmers, gets useful things done easily, and is open source then compensation is usually lower. But it is likely the best  framework to adopt if you start your own company.

> > By the way, you should consider editing your SO question to make it clear to other users.
> 
> All right, I'll change the title of the question.

Great.

**Comment by GibranLara** [comment]
I already saw the video you recommended me.

I also started working with Vue.

**Comment by andre-artus** [comment]
Hi Gibrán,

Can you tell we what tool did you use to make the Entity Relationship Diagram (ERD)?

**Comment by GibranLara** [comment]
> Can you tell we what tool did you use to make the Entity Relationship Diagram (ERD)?

Yes Andre! I used Star UML. [Link](http://staruml.io/)

**Comment by andre-artus** [comment]
Thanks Gibrán.

**Comment by GibranLara** [comment]
Sorry Andre, do you have experience or have you worked with Vue 2.0>?

**Comment by andre-artus** [comment]
A little bit, do you have a question?

**Comment by GibranLara** [comment]
Yes. Currently in my front end I have 3 components.

- Proyectos.vue
- Reuniones.vue
- Reunion.vue

Where:

`Proyecto`: it is the main component. It is where the list of projects is shown.
`Reuniones`: it is the component where the list of meetings of each project is shown.
`Reunion`: is the component where the data is shown a meeting, in addition to the list of participants.
 
My problem is that I do not know how to be passing a 'proyecto' object between components. I explain:

![image](https://user-images.githubusercontent.com/22408615/49321011-8cf15f00-f4ca-11e8-904c-b0969d9b422c.png)


My interaction logic that I want is:

1. In my Proyectos component I do the GET of all the projects. This array of objects allow create a table.

2. From the previous list I want to select a project and call the Reuniones component. In this Reuniones component I want to tabulate the meetings.

3. From the list of meetings I want to select a meeting and call the Reunion component. In this component Reunion I want to show the data of the meeting and tabulate the participants.

Proyecto
![Proyecto](https://user-images.githubusercontent.com/22408615/49321030-a7c3d380-f4ca-11e8-91d6-461702413892.png)

Meetings of the proyecto
![image](https://user-images.githubusercontent.com/22408615/49321055-c2964800-f4ca-11e8-963d-5c37fe1d2ed0.png)

Detail of the meeting 
![image](https://user-images.githubusercontent.com/22408615/49321066-e194da00-f4ca-11e8-8673-b349ef00c47d.png)


What I am looking for is that during the transition in the components a `proyecto` object is being edited, and when saving changes in any component, to send the put/post of the complete `proyecto` object. I want to avoid passing the `id` to a component and that in this I have to do a `get` again to recover it, because the object of the proyecto itself already has it since the first `getAll`. Something I'm also trying to avoid is to use `params`, not send data by the url, if not all keep it between components.

 I had read something about a global bus and events, but I can not quite understand it. I also read something from Vuex, but I do not understand much either.

**Comment by andre-artus** [comment]
There are, to my knowledge, 3 different ways of doing state management in the way you want.
* custom events.
* Event Bus
* Vuex

While your app is small any of these would work just fine.  Vuex is what I would gravitate towards, as it has some advantages that the others don't. But as this is a learning project it is well worth trying all three. 

If you want to put up a basic app on GH that I can collaborate on, I can help you there. If you do, can I ask that you please give your names and comments in English (or provide translations)? My keyboard is not set up for easy use of diacritical marks as they are not often used by me, so that, along with translating, slows things down for me.

I already know the basic translations of the primary objects:

Proyecto -> Project
Reuni'on -> Meeting
Participantes -> Participants
Nombre -> Name
Area -> Location
Actiones -> Actions
Fecha -> Date+Time

**Comment by andre-artus** [comment]
Some information for custom events:

https://medium.com/introcept-hub/you-will-be-surprised-custom-event-in-vue-js-c713afdff8e5
https://vuejs.org/v2/guide/components-custom-events.html

**Comment by andre-artus** [comment]
https://vuejs.org/v2/guide/components.html#Passing-Data-to-Child-Components-with-Props
https://vuejs.org/v2/guide/components.html#Sending-Messages-to-Parents-with-Events

These are the specifics, but read the whole section. 

Also Vue.js has good free videos.

Send me the link if you want me to collaborate on your repo.

**Comment by GibranLara** [comment]
> If you want to put up a basic app on GH that I can collaborate on, I can help you there.

It seems perfect to me, I would be grateful too. 

> Send me the link if you want me to collaborate on your repo.

Currently I already have the frontend project up in GH. Do you want us to work on that or do I create a new empty project?

**Comment by andre-artus** [comment]
> Currently I already have the frontend project up in GH. Do you want us to work on that or do I create a new empty project?

Using the existing one is fine by me. Could you just add some translations (in comments)?

**Comment by GibranLara** [comment]
> Using the existing one is fine by me. Could you just add some translations (in comments)?

Very well, understood. Later I will notify you when I already have the code commenting in English.

**Comment by GibranLara** [comment]
> > Using the existing one is fine by me. Could you just add some translations (in comments)?

Ready Andre.

Actually from the code I do not have many different words to the list that you point out. I have:

- Cancelar (Cancel), 
- Guardar (Save)
- Aceptar (Accept)
- Buscar (Search) 
- Borrar (Delete)
- Nuevo (New)

Frontend repository: https://github.com/GibranLara/vue-front
Backend repository: https://github.com/GibranLara/spring-back

How would you like us to be communicating and collaborating?

**Comment by andre-artus** [comment]
Cool. Maybe next time I visit a Spanish speaking country  I'll know more than "dos cervezas, por favor". :)
